lib/spack/spack/binary_distribution.py

```diff
@@ -645,3 +681,131 @@ def get_keys(install=False, trust=False, force=False):
                 else:
                     tty.msg('Will not add this key to trusted keys.'
                             'Use -t to install all downloaded keys')
+
+
+def needs_rebuild(spec, mirror_url, rebuild_on_errors=False):
+    if not spec.concrete:
+        raise ValueError('spec must be concrete to check against mirror')
+
+    pkg_name = spec.name
+    pkg_version = spec.version
+
+    pkg_hash = spec.dag_hash()
+    pkg_full_hash = spec.full_hash()
+
+    tty.debug('Checking {0}-{1}, dag_hash = {2}, full_hash = {3}'.format(
+        pkg_name, pkg_version, pkg_hash, pkg_full_hash))
+    tty.debug(spec.tree())
+
+    # Try to retrieve the .spec.yaml directly, based on the known
+    # format of the name, in order to determine if the package
+    # needs to be rebuilt.
+    build_cache_dir = build_cache_directory(mirror_url)
+    spec_yaml_file_name = tarball_name(spec, '.spec.yaml')
+    file_path = os.path.join(build_cache_dir, spec_yaml_file_name)
+
+    result_of_error = 'Package ({0}) will {1}be rebuilt'.format(
```

Comment:
If you don't like how the code looks, I'm happy to change it.  However, the string I format with may be the empty string or else it may be `"not "`.   So there shouldn't be a missing space in the log/error message.

Reply:
Got it - I overlooked that.